### ISSUE TITLE
IOS XR: AttributeError: 'CiscoXrSSH' object has no attribute 'receive_data_generator'

### ISSUE THREAD

**Issue by rlanyi** [napalm-automation/napalm#1312]
### Description of Issue

Calling Cisco ASR9K devices (IOS XR) via SSH from Napalm CLI is unreliable, seemingly only very small responses can be gathered. With most calls I get `AttributeError: 'CiscoXrSSH' object has no attribute 'receive_data_generator'` error, sometimes during the built-in `get_facts` call, sometimes during the call passed on the command line. Even if my command is also `get_facts` it happens that the built-in one runs without issues but the one provided by me fails. Sometimes it fails during the built-in call. It can never show the response of larger calls like `get_bgp_neighbors`.

### Did you follow the steps from https://github.com/napalm-automation/napalm#faq

- [x] Yes
- [ ] No

### napalm version
I'm running napalm from the Dockerized version of [peering-manager](https://github.com/peering-manager/peering-manager/) v1.2.1 .

```
napalm==3.1.0
```

### Network operating system version

```
Cisco IOS XR Software, Version 6.1.2[Default]
```

#### 1st try: built-in `get_facts` call runs properly, `get_facts` passed via the command line fails
<details><summary>Click for details</summary>
<p>

```
root@1dab935bca2b:/opt/peering-manager# napalm --debug --user root --password xxxx --vendor iosxr --optional_args "ssh_config_file='/root/.ssh/config'" router1 call get_facts
2020-10-28 20:41:43,221 - napalm - DEBUG - Starting napalm's debugging tool
2020-10-28 20:41:43,221 - napalm - DEBUG - Gathering napalm packages
2020-10-28 20:41:43,221 - napalm - DEBUG - napalm==3.1.0
2020-10-28 20:41:43,221 - napalm - DEBUG - get_network_driver - Calling with args: ('iosxr',), {}
2020-10-28 20:41:43,221 - napalm - DEBUG - get_network_driver - Successful
2020-10-28 20:41:43,221 - napalm - DEBUG - __init__ - Calling with args: (<class 'napalm.iosxr.iosxr.IOSXRDriver'>, 'router1', 'root'), {'password': '*******', 'timeout': 60, 'optional_args': {'ssh_config_file': '/root/.ssh/config'}}
2020-10-28 20:41:43,222 - napalm - DEBUG - __init__ - Successful
2020-10-28 20:41:43,222 - napalm - DEBUG - pre_connection_tests - Calling with args: (<napalm.iosxr.iosxr.IOSXRDriver object at 0x7f8c067282e0>,), {}
2020-10-28 20:41:43,222 - napalm - DEBUG - open - Calling with args: (<napalm.iosxr.iosxr.IOSXRDriver object at 0x7f8c067282e0>,), {}
2020-10-28 20:41:46,145 - napalm - DEBUG - open - Successful
2020-10-28 20:41:46,145 - napalm - DEBUG - connection_tests - Calling with args: (<napalm.iosxr.iosxr.IOSXRDriver object at 0x7f8c067282e0>,), {}
2020-10-28 20:41:46,146 - napalm - DEBUG - get_facts - Calling with args: (<napalm.iosxr.iosxr.IOSXRDriver object at 0x7f8c067282e0>,), {}
2020-10-28 20:41:57,415 - napalm.pyIOSXR.iosxr - DEBUG - <Element Response at 0x7f8c0670e440>
2020-10-28 20:42:00,274 - napalm.pyIOSXR.iosxr - DEBUG - <Element Response at 0x7f8c0672c400>
2020-10-28 20:42:00,301 - napalm - DEBUG - Gathered facts:
{
    "vendor": "Cisco",
    "os_version": "6.1.2",
    "hostname": "router1",
    "uptime": 19475078,
    "serial_number": "FOX24000",
    "fqdn": "router1",
    "model": "ASR-9006-AC-V2",
    "interface_list": [
        "BVI1",
        "Bundle-Ether1",
        "GigabitEthernet0/1/0/0",
(...)
    ]
}
{
    "vendor": "Cisco",
    "os_version": "6.1.2",
    "hostname": "router1",
    "uptime": 19475078,
    "serial_number": "FOX24000",
    "fqdn": "router1",
    "model": "ASR-9006-AC-V2",
    "interface_list": [
        "BVI1",
        "Bundle-Ether1",
        "GigabitEthernet0/1/0/0",
(...)
    ]
}
2020-10-28 20:42:00,301 - napalm - DEBUG - get_facts - Successful
2020-10-28 20:42:00,301 - napalm - DEBUG - method - Calling with args: (<napalm.iosxr.iosxr.IOSXRDriver object at 0x7f8c067282e0>, 'get_facts'), {}
2020-10-28 20:42:00,301 - napalm - DEBUG - get_facts - Attempting to resolve method
2020-10-28 20:42:00,301 - napalm - DEBUG - get_facts - Attempting to call method with kwargs: {}
2020-10-28 20:42:12,707 - napalm - ERROR - method - Failed: 'CiscoXrSSH' object has no attribute 'receive_data_generator'

================= Traceback =================

Traceback (most recent call last):
  File "/usr/local/bin/napalm", line 8, in <module>
    sys.exit(main())
  File "/usr/local/lib/python3.9/site-packages/napalm/base/clitools/cl_napalm.py", line 308, in main
    run_tests(args)
  File "/usr/local/lib/python3.9/site-packages/napalm/base/clitools/cl_napalm.py", line 291, in run_tests
    call_getter(device, args.method, **method_kwargs)
  File "/usr/local/lib/python3.9/site-packages/napalm/base/clitools/cl_napalm.py", line 27, in wrapper
    r = func(*args, **kwargs)
  File "/usr/local/lib/python3.9/site-packages/napalm/base/clitools/cl_napalm.py", line 255, in call_getter
    r = func(**kwargs)
  File "/usr/local/lib/python3.9/site-packages/napalm/iosxr/iosxr.py", line 166, in get_facts
    facts_rpc_reply = ETREE.fromstring(self.device.make_rpc_call(facts_rpc_request))
  File "/usr/local/lib/python3.9/site-packages/napalm/pyIOSXR/iosxr.py", line 172, in make_rpc_call
    result = self._execute_rpc(rpc_command)
  File "/usr/local/lib/python3.9/site-packages/napalm/pyIOSXR/iosxr.py", line 426, in _execute_rpc
    response = self._send_command(xml_rpc_command, delay_factor=delay_factor)
  File "/usr/local/lib/python3.9/site-packages/napalm/pyIOSXR/iosxr.py", line 399, in _send_command
    return self._send_command(
  File "/usr/local/lib/python3.9/site-packages/napalm/pyIOSXR/iosxr.py", line 351, in _send_command
    output += self._netmiko_recv()  # try to read some more
  File "/usr/local/lib/python3.9/site-packages/napalm/pyIOSXR/iosxr.py", line 412, in _netmiko_recv
    for tmp_output in self.device.receive_data_generator():
AttributeError: 'CiscoXrSSH' object has no attribute 'receive_data_generator'
```
</p>
</details>

#### 2nd try: built-in `get_facts` call fails
<details><summary>Click for details</summary>
<p>

```
root@1dab935bca2b:/opt/peering-manager# napalm --debug --user root --password xxxx --vendor iosxr --optional_args "ssh_config_file='/root/.ssh/config'" router1 call get_facts
2020-10-28 20:45:18,013 - napalm - DEBUG - Starting napalm's debugging tool
2020-10-28 20:45:18,013 - napalm - DEBUG - Gathering napalm packages
2020-10-28 20:45:18,013 - napalm - DEBUG - napalm==3.1.0
2020-10-28 20:45:18,013 - napalm - DEBUG - get_network_driver - Calling with args: ('iosxr',), {}
2020-10-28 20:45:18,014 - napalm - DEBUG - get_network_driver - Successful
2020-10-28 20:45:18,014 - napalm - DEBUG - __init__ - Calling with args: (<class 'napalm.iosxr.iosxr.IOSXRDriver'>, 'router1', 'root'), {'password': '*******', 'timeout': 60, 'optional_args': {'ssh_config_file': '/root/.ssh/config'}}
2020-10-28 20:45:18,014 - napalm - DEBUG - __init__ - Successful
2020-10-28 20:45:18,014 - napalm - DEBUG - pre_connection_tests - Calling with args: (<napalm.iosxr.iosxr.IOSXRDriver object at 0x7f5587dbb2e0>,), {}
2020-10-28 20:45:18,014 - napalm - DEBUG - open - Calling with args: (<napalm.iosxr.iosxr.IOSXRDriver object at 0x7f5587dbb2e0>,), {}
2020-10-28 20:45:21,070 - napalm - DEBUG - open - Successful
2020-10-28 20:45:21,071 - napalm - DEBUG - connection_tests - Calling with args: (<napalm.iosxr.iosxr.IOSXRDriver object at 0x7f5587dbb2e0>,), {}
2020-10-28 20:45:21,071 - napalm - DEBUG - get_facts - Calling with args: (<napalm.iosxr.iosxr.IOSXRDriver object at 0x7f5587dbb2e0>,), {}
2020-10-28 20:45:33,378 - napalm - ERROR - get_facts - Failed: 'CiscoXrSSH' object has no attribute 'receive_data_generator'

================= Traceback =================

Traceback (most recent call last):
  File "/usr/local/bin/napalm", line 8, in <module>
    sys.exit(main())
  File "/usr/local/lib/python3.9/site-packages/napalm/base/clitools/cl_napalm.py", line 308, in main
    run_tests(args)
  File "/usr/local/lib/python3.9/site-packages/napalm/base/clitools/cl_napalm.py", line 287, in run_tests
    call_facts(device)
  File "/usr/local/lib/python3.9/site-packages/napalm/base/clitools/cl_napalm.py", line 27, in wrapper
    r = func(*args, **kwargs)
  File "/usr/local/lib/python3.9/site-packages/napalm/base/clitools/cl_napalm.py", line 195, in call_facts
    facts = device.get_facts()
  File "/usr/local/lib/python3.9/site-packages/napalm/iosxr/iosxr.py", line 166, in get_facts
    facts_rpc_reply = ETREE.fromstring(self.device.make_rpc_call(facts_rpc_request))
  File "/usr/local/lib/python3.9/site-packages/napalm/pyIOSXR/iosxr.py", line 172, in make_rpc_call
    result = self._execute_rpc(rpc_command)
  File "/usr/local/lib/python3.9/site-packages/napalm/pyIOSXR/iosxr.py", line 426, in _execute_rpc
    response = self._send_command(xml_rpc_command, delay_factor=delay_factor)
  File "/usr/local/lib/python3.9/site-packages/napalm/pyIOSXR/iosxr.py", line 399, in _send_command
    return self._send_command(
  File "/usr/local/lib/python3.9/site-packages/napalm/pyIOSXR/iosxr.py", line 351, in _send_command
    output += self._netmiko_recv()  # try to read some more
  File "/usr/local/lib/python3.9/site-packages/napalm/pyIOSXR/iosxr.py", line 412, in _netmiko_recv
    for tmp_output in self.device.receive_data_generator():
AttributeError: 'CiscoXrSSH' object has no attribute 'receive_data_generator'
```
</p>
</details>

#### 3rd try: built-in `get_facts` call runs properly, `get_facts` passed via the command runs properly
<details><summary>Click for details</summary>
<p>

```
root@1dab935bca2b:/opt/peering-manager# napalm --debug --user root --password xxxx --vendor iosxr --optional_args "ssh_config_file='/root/.ssh/config'" router1 call get_facts
2020-10-28 20:47:25,884 - napalm - DEBUG - Starting napalm's debugging tool
2020-10-28 20:47:25,884 - napalm - DEBUG - Gathering napalm packages
2020-10-28 20:47:25,884 - napalm - DEBUG - napalm==3.1.0
2020-10-28 20:47:25,884 - napalm - DEBUG - get_network_driver - Calling with args: ('iosxr',), {}
2020-10-28 20:47:25,885 - napalm - DEBUG - get_network_driver - Successful
2020-10-28 20:47:25,885 - napalm - DEBUG - __init__ - Calling with args: (<class 'napalm.iosxr.iosxr.IOSXRDriver'>, 'router1', 'root'), {'password': '*******', 'timeout': 60, 'optional_args': {'ssh_config_file': '/root/.ssh/config'}}
2020-10-28 20:47:25,885 - napalm - DEBUG - __init__ - Successful
2020-10-28 20:47:25,885 - napalm - DEBUG - pre_connection_tests - Calling with args: (<napalm.iosxr.iosxr.IOSXRDriver object at 0x7f3a91ee92e0>,), {}
2020-10-28 20:47:25,885 - napalm - DEBUG - open - Calling with args: (<napalm.iosxr.iosxr.IOSXRDriver object at 0x7f3a91ee92e0>,), {}
2020-10-28 20:47:28,914 - napalm - DEBUG - open - Successful
2020-10-28 20:47:28,914 - napalm - DEBUG - connection_tests - Calling with args: (<napalm.iosxr.iosxr.IOSXRDriver object at 0x7f3a91ee92e0>,), {}
2020-10-28 20:47:28,914 - napalm - DEBUG - get_facts - Calling with args: (<napalm.iosxr.iosxr.IOSXRDriver object at 0x7f3a91ee92e0>,), {}
2020-10-28 20:47:39,967 - napalm.pyIOSXR.iosxr - DEBUG - <Element Response at 0x7f3a91edccc0>
2020-10-28 20:47:42,380 - napalm.pyIOSXR.iosxr - DEBUG - <Element Response at 0x7f3a91ea4880>
2020-10-28 20:47:42,417 - napalm - DEBUG - Gathered facts:
{
    "vendor": "Cisco",
    "os_version": "6.1.2",
    "hostname": "router1",
    "uptime": 19475420,
    "serial_number": "FOX24000",
    "fqdn": "router1",
    "model": "ASR-9006-AC-V2",
    "interface_list": [
        "BVI1",
        "Bundle-Ether1",
        "GigabitEthernet0/1/0/0",
(...)
    ]
}
{
    "vendor": "Cisco",
    "os_version": "6.1.2",
    "hostname": "router1",
    "uptime": 19475420,
    "serial_number": "FOX24000",
    "fqdn": "router1",
    "model": "ASR-9006-AC-V2",
    "interface_list": [
        "BVI1",
        "Bundle-Ether1",
        "GigabitEthernet0/1/0/0",
(...)
    ]
}
2020-10-28 20:47:42,417 - napalm - DEBUG - get_facts - Successful
2020-10-28 20:47:42,417 - napalm - DEBUG - method - Calling with args: (<napalm.iosxr.iosxr.IOSXRDriver object at 0x7f3a91ee92e0>, 'get_facts'), {}
2020-10-28 20:47:42,417 - napalm - DEBUG - get_facts - Attempting to resolve method
2020-10-28 20:47:42,417 - napalm - DEBUG - get_facts - Attempting to call method with kwargs: {}
2020-10-28 20:47:54,129 - napalm.pyIOSXR.iosxr - DEBUG - <Element Response at 0x7f3a91edc380>
2020-10-28 20:47:56,620 - napalm.pyIOSXR.iosxr - DEBUG - <Element Response at 0x7f3a91ed9e00>
2020-10-28 20:47:56,648 - napalm - DEBUG - get_facts - Response
{
    "vendor": "Cisco",
    "os_version": "6.1.2",
    "hostname": "router1",
    "uptime": 19475434,
    "serial_number": "FOX24000",
    "fqdn": "router1",
    "model": "ASR-9006-AC-V2",
    "interface_list": [
        "BVI1",
        "Bundle-Ether1",
        "GigabitEthernet0/1/0/0",
(...)
    ]
}
2020-10-28 20:47:56,650 - napalm - DEBUG - method - Successful
2020-10-28 20:47:56,650 - napalm - DEBUG - close - Calling with args: (<napalm.iosxr.iosxr.IOSXRDriver object at 0x7f3a91ee92e0>,), {}
2020-10-28 20:47:56,650 - napalm.iosxr.iosxr - DEBUG - Closed connection with device router1
2020-10-28 20:47:56,650 - napalm - DEBUG - close - Successful
2020-10-28 20:47:56,650 - napalm - DEBUG - post_connection_tests - Calling with args: (<napalm.iosxr.iosxr.IOSXRDriver object at 0x7f3a91ee92e0>,), {}
2020-10-28 20:47:56,650 - napalm.iosxr.iosxr - DEBUG - Closed connection with device router1
```
</p>
</details>

### Additional info

In all examples above this is the SSH config file.

```
Host router1
HostName 192.168.0.10
User root
KexAlgorithms diffie-hellman-group14-sha1,diffie-hellman-group1-sha1
Ciphers aes256-ctr
```

This is the XML TTY part of my IOS XR config. I've tried to use `iteration on size nnn` with different values but I couldn't see any improvement.

```
xml agent tty
 iteration off
!
```

**Comment by mirceaulinic** [comment]
Hi @rlanyi. Could you share what Netmiko version are you using?

**Comment by rlanyi** [comment]
Hi @mirceaulinic, thanks for looking into this. Netmiko version is latest, `3.3.2`.

```
# cat /usr/local/lib/python3.9/site-packages/netmiko*/METADATA | grep '^Version'
Version: 3.3.2
```

**Comment by Tristou27** [comment]
Hello.

I've the same behavior here, with IOS-XR 6.5.3. It's not happening that much but I've seen it 3 times today.
Using napalm 3.2.0, netmiko 3.3.2

For information, after some researches I found that the function receive_data_generator() has been removed from the netmiko BaseConnection class since 2016, see commit [d93a643](https://github.com/ktbyers/netmiko/commit/d93a6435e67a37ec78557d77918f7adff6cf8456)

An update on pyIOSXR could solve this but I'm not sure to understand the initial issue solved by the call to receive_data_generator.

**Comment by mirceaulinic** [comment]
Could any of you give https://github.com/napalm-automation/napalm/pull/1332 a try and confirm it works well?

**Comment by rlanyi** [comment]
Hi @ktbyers , I've tested this in the same (dockerized) environment as above and out of 10 tries I've never received the `receive_data_generator` error anymore. I've other errors but I think these are unrelated to this issue so I think the change can be merged.

For reference, I've the following errors, in 60% of the cases if I run the command above and in 40% of the cases if I add `--optional_args "fast_cli=True"`:

- `lxml.etree.XMLSyntaxError: Start tag expected, '<' not found, line 1, column 1`
- `napalm.pyIOSXR.exceptions.InvalidXMLResponse: Unable to process the XML Response from the device!`
- `raise XMLCLIError(output.strip(), self)`

**Comment by mirceaulinic** [comment]
Solved via #1332, as @rlanyi confirmed in #1342.
